### ISSUE TITLE
Create new project files daily

### ISSUE THREAD

**Issue by AndrewM-** [geopaparazzi/geopaparazzi#529]
Garmin GPS units can create new data files daily.  I would appreciate if there was an option to create a new data file each day in geopaparazzi.  This would allow me to integrate gpap files into my existing data management system.  My system is designed to be capable of operation both as a filing system and as a database driven system.  It was also designed to make conflicts impossible and to survive the loss of all management software.  Non-profit organisations are subject to stop and start funding and records have to be as indestructible as possible.  This is my solution.

![menustructure](https://user-images.githubusercontent.com/5060805/41814360-2f781ebe-778d-11e8-8c57-d74783bfd0aa.jpg)

All data from one date is stored in a folder and data from different devices in stored in subfolders which are named using the devices serial number.  In the case of cameras, I used the serial number of the SD card as cameras do not write their serial number into the photos exif data by default.  This works for me as my organisation will have lots of cameras, gps units and other devices which are the same make and model.  If I could get daily gpap files and the serial number of the device that created them, then I could use my existing approach and tool chain to manage them.  The filing system can by created manually, but I have automated the process so when people connect a device and run my data management application, the device is scanned and the new files automatically downloaded, registered and filed correctly.  The html page is generated automatically and list the contents of the subfolders.

Gpap data is both file data and database data and I think that it is important that it works in both worlds.
The primary need for daily data files is facilitate efficient backup of each day's data.  The secondary need if for each guy to be able to see what he has done each day.  If the guys swap devices, then they end up with someone else's data and when they download and view the data they get confused.  They asked me to find a way to resolve this issue.  Automated creation of a new project file each day is possibly the simplest solution.

**Comment by brentfraser** [comment]
Interesting stuff, but I'm a little confused as to your current workflow.  Is it:
        * there are many data collection users and devices
        * the collection project lasts many days
    1.  Every morning a user gets a device and creates a new [daily] Project file using Geopaparazzi
    2.  The user collects data during the day into the new daily Project file
    3.  At the end of the day the user connects (via USB?) and transfers the daily Project file to the server
    4.  The users then deletes the daily Project file from the device.

Problem:
    There is no easy way (currently) to ID the device when uploading the daily Project file.

The Project files may be created on the device, but they may be created on the server and shipped to many devices so it would be difficult to set a device ID on the server.

Possible Solution:
  Add a "lastdevice" row to the Project metadata table with a value of the Android serial number (or IMEI).  It would get set by Geopaparazzi when the Project is opened.  Your server could upload the Project file to a temp dir, get the value and use it to create the appropriate folder.

Your thoughts?

**Comment by AndrewM-** [comment]
I need to get a daily backup of data from the device going.  That protects us from loss of data due to loss or damage to the device.  Hence the need for daily backups.  With some projects, there are a lot of photos, for example fire monitoring plots.  Each plot has 12 photos and 8 plots might be recorded each day.  Without making a new file daily, the size of the gpap files will rapidly become difficult to manage.

You are correct about the workflow, except the user retains the daily project files on the device until the end of the field survey, which maybe two weeks.  When the data is retained on the device, the device acts as an additional backup set.  We have many guys in the field for many days and each team will have a device of the same make and model.  That means I could be getting from 4 to 10 daily backups.  

I think that adding the device serial number is only important when the gpap file is created in the device.  The android serial number would be fine as the serial number plus the time stamp would make a GUID.

I used to use MD5 hashes and use them for a GUID but if there is corruption or data loss somewhere, I can't recreate the same hash.  When I use the serial number, if my computer dies, I can plug the devices into another computer and create exactly the same set of project folders.  This ability to recover from hardware and software failures is very important.  Hence I use a filing system which can be rebuilt from any surviving data sources.

The only issue I can see is if people copy the data file to another device after it has been created and then use the original device in parallel.  That would be unlikely.

Currently, I am heading toward using gpap files in the same way geopackages would be used in QGIS.  Each day in a separate file makes life easy.

**Comment by brentfraser** [comment]
The proposed "lastdevice" row to the Project metadata table would be updated whenever the Project database file is opened by Geopaparazzi, not just at Project database creation time so it would handle the case of creating the Project database on one device (or a server) and distributing it to many devices.

It would not handle the case of collecting data with one device then moving/copying to another device and collecting more data.  (Maybe we could put an entry in the log table to handle that rare case).
----

 We are in the process of implementing Cloud Profiles to handle this kind of campaign workflow.  The latest version of Geopaparazzi has Import -> Cloud Profiles to transfer Project and Spatialite files to Geopaparazzi.   I'm currently implementing Export -> Cloud Profiles to handle the upload of Project and Spatialite files to a server.  We are implementing the "reference" server software as well:
https://groups.google.com/forum/#!topic/geopaparazzi-users/umjOK0TQjD0
https://geoanalytic.github.io/a-reference-server-for-geopaparazzi-cloud-profiles-part-2/
https://github.com/geoanalytic/cookiecutter-geopaparazzi-server

**Comment by AndrewM-** [comment]
I certainly try to think stuff through before asking for something here.  The geopaparazzi practice of prefixing the gpap file with a timestamp can be used to facilitate storing info without file name collisions and is very easy to implement in a data management application.  Because of my time constrains, I will go this way but it creates the issue of not knowing which team created the data by looking at the file name.  If I make a data management application, I can require that the team name be entered and add a folder with the team ID, then put the geopaparazzi image folder and gpap file in the team folder.  That would work.  The data management application would just create folders in the correct place for new incoming gpap files and collect some metadata on the gpap files probably from the internal gpap metadata and from comments entered by the user.  The device make, model and serial number still has utility as it can help track issues with devices and also some android devices are special capabilities such as underwater cameras.  So I still like the idea about having info on which device the gpap file was in.

With respect to cloud profiles, the mbtiles layers that I use are 3GB in size.  I should always put these on the device using a cable to avoid pulling them in over a cell phone network.  A lot of our tablets also lack cell phone network access.   I hope that the cloud profiles will check for files before pulling new ones in.  

I am heading toward replacing my ms-access based solutions with django/python and postGIS.  That is the magic combination for QGIS users.

**Comment by brentfraser** [comment]
I agree that the device name, serial number, etc would be useful info for Geopaparazzi to track.  I was just trying to think of short-term solutions until we get around to adding the enhancement to Geopaparazzi.  I've been looking into getting the device serial number but it seems it is not available on all devices, and it might not be easy of the user to use it to identify the physical device.  The device name might be easier for the user, but it might not be unique since it is settable by the user.

**Comment by AndrewM-** [comment]
I built a short term solution in ms-access that facilitates daily backups to a laptop in the field or directly to the office server.  It works like this and it works off-line.  File transfer from device is by cable and uses the emulated storage (MTP) based folder for the device as displayed in Windows.
![2018-07-04_dataregister](https://user-images.githubusercontent.com/5060805/42249552-362af634-7f6f-11e8-9c56-c23d6e6f418c.jpg)
This approach deals with my off-line back-up requirements.  It first backs up the gpap file onto the local computer and registers the Gpap file.  If the office network is available, then files can be moved from the holding area on the local PC to an intelligent filing system on the server.  There is a lot of validation code under the surface to prevent users from making silly mistakes.  The Geopaparazzi Server was released on day after I finished the ms-access application.  I think the approaches are complementary.

**Comment by moovida** [comment]
Your application looks stunning @AndrewM- . The GSS might be similar, but at the currnt time it doesn't have particular filter using time. 

Anyways, I though I will give this issue a try once I have time to do fixes again. It shouldn't be too difficult to tell geopaparazzi to create a new project for the day once geopaparazzi opens. This might be a preference. Once chosen, at startup, if the day matches and existing gpap, open it, else create it.

**Comment by AndrewM-** [comment]
I agree, that is how daily files should be implemented.
